code/parachain/frame/pablo/src/dual_asset_constant_product.rs

```diff
@@ -143,42 +143,29 @@ impl<T: Config> DualAssetConstantProduct<T> {
 		Ok((*first_asset_amount, *second_asset_amount, lp_issued.safe_sub(&lp_amount)?))
 	}
 
-	pub(crate) fn do_swap(
+	pub(crate) fn get_exchange_value(
 		pool: &BasicPoolInfo<T::AccountId, T::AssetId, ConstU32<2>>,
 		pool_account: &T::AccountId,
 		in_asset: AssetAmount<T::AssetId, T::Balance>,
-		min_receive: AssetAmount<T::AssetId, T::Balance>,
-		apply_fees: bool,
-	) -> Result<(T::Balance, T::Balance, Fee<T::AssetId, T::Balance>), DispatchError> {
+		out_asset_id: T::AssetId,
+		_apply_fees: bool,
+	) -> Result<
+		(AssetAmount<T::AssetId, T::Balance>, AssetAmount<T::AssetId, T::Balance>),
+		DispatchError,
+	> {
 		let pool_assets = Self::get_pool_balances(pool, pool_account);
-		let base_asset = pool_assets.get(&min_receive.asset_id).ok_or(Error::<T>::PairMismatch)?;
-		let quote_asset = pool_assets.get(&in_asset.asset_id).ok_or(Error::<T>::PairMismatch)?;
-		ensure!(base_asset.1 > 0 && quote_asset.1 > 0, Error::<T>::NotEnoughLiquidity);
+		let a_sent = T::Convert::convert(in_asset.amount);
+		let (w_i, b_i) = pool_assets.get(&in_asset.asset_id).ok_or(Error::<T>::PairMismatch)?;
+		let (w_o, b_o) = pool_assets.get(&out_asset_id).ok_or(Error::<T>::PairMismatch)?;
 
-		// TODO (vim): Fees refactored later
-		let fee = if apply_fees {
-			pool.fee_config.calculate_fees(in_asset.asset_id, in_asset.amount)
-		} else {
-			Fee::<T::AssetId, T::Balance>::zero(in_asset.asset_id)
-		};
-		// Charging fees "on the way in"
-		// https://balancer.gitbook.io/balancer/core-concepts/protocol/index#out-given-in
-		let quote_amount_excluding_lp_fee =
-			T::Convert::convert(in_asset.amount.safe_sub(&fee.fee)?);
-		let base_amount = compute_out_given_in(
-			quote_asset.0,
-			base_asset.0,
-			quote_asset.1,
-			base_asset.1,
-			quote_amount_excluding_lp_fee,
-		)?;
-		ensure!(base_amount > 0 && quote_amount_excluding_lp_fee > 0, Error::<T>::InvalidAmount);
+		// TODO: Provide fee calculations or value
+		let amm_pair =
+			compute_out_given_in_new::<_>(*w_i, *w_o, *b_i, *b_o, a_sent, Permill::zero())?;
```

Comment:
Why not provide the fee value now? You can take it from the pool data structure

Reply:
Misunderstood the previous TODO. Has been addressed